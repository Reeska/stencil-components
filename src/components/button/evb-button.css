:host {
  overflow: hidden;
  text-decoration: none;
  text-overflow: ellipsis;
  white-space: nowrap;
}

:host([disabled]) {
  pointer-events: none;
}
:host([type=icon]) {
  height: calc(2.5em - 0px);
  width: calc(2.5em - 0px);
}
::slotted(evb-icon) {
  height: calc(100% - 1rem);
  width: calc(100% - 1rem);
  padding-top: .25rem;
  fill: currentColor;
}
button[type=icon] {
  padding: 0;
  width: 100%;
  height: 100%;
}

button, a {
  border: 1px solid transparent;
  color: var(--evb-text-color-invert);
  font-size: 1em;
  font-weight: 500;
  line-height: 1.5;
  padding: calc(.5rem - 1px) calc(1rem - 1px); /* minus 1px to compensate for the border */
  transition: all .1s ease-in;
}
button:hover, a:hover {
  transition: all .2s ease-in;
}

button {
  background-color: var(--evb-primary-color);
}
button:hover {
  background-color: var(--evb-primary-color-darker);
}
button.secondary {
  background-color: var(--evb-secondary-color);
}
button.secondary:hover {
  background-color: var(--evb-secondary-color-darker);
}
button.danger {
  background-color: var(--evb-danger-color);
}
button.danger:hover {
  background-color: var(--evb-danger-color-darker);
}

button[disabled] {
  opacity: .6;
  pointer-events: none;
}
button.link {
  background-color: transparent;
  color: var(--evb-primary-color);
  cursor: pointer;
  text-decoration: none;
}
button.link:hover {
  background-color: transparent;
  text-decoration: underline;
}
button.secondary.link {
  color: var(--evb-secondary-color);
}
button.danger.link {
  color: var(--evb-danger-color);
}
button.ghost {
  background-color: transparent;
  color: var(--evb-primary-color);
  border-color: currentColor;
}
button.ghost:hover {
  color: var(--evb-text-color-invert);
  background-color: var(--evb-primary-color-darker);
}
button.secondary.ghost {
  color: var(--evb-secondary-color);
}
button.secondary.ghost:hover {
  background-color: var(--evb-secondary-color);
  color: var(--evb-text-color-invert);
}
button.danger.ghost {
  color: var(--evb-danger-color);
}
button.danger.ghost:hover {
  background-color: var(--evb-danger-color);
  color: var(--evb-text-color-invert);
}
button.pill {
  border-radius: 2rem;
}

a {
  color: var(--evb-primary-color);
  text-decoration: none;
}
a:hover {
  text-decoration: underline;
}

a.secondary {
  color: var(--evb-secondary-color);
}
a.danger {
  color: var(--evb-danger-color);
}
