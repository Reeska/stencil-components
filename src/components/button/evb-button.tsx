import { Component, ComponentInterface, Event, EventEmitter, Prop } from '@stencil/core';

/**
 * Button component with click, focus and blur support
 */
@Component({
  tag: 'evb-button',
  styleUrl: 'evb-button.css',
  shadow: true
})
export class EvbButton implements ComponentInterface {

  /** The type of the button, leave empty in case of a link */
  @Prop() type?: 'submit' | 'button' | 'reset' = 'button';
  /** The target of the anchor tag */
  @Prop() href?: string;
  /** Invert the coloring of the button */
  @Prop() ghost = false;
  /** Extra rounded colors */
  @Prop() pill = false;

  /** Whether or not the button is disabled */
  @Prop({ reflectToAttr: true }) disabled = false;

  /** Focus event */
  @Event() evbFocus!: EventEmitter<void>;
  /** Blur event */
  @Event() evbBlur!: EventEmitter<void>;

  private onBlur = () => this.evbBlur.emit();
  private onFocus = () => this.evbFocus.emit();

  /**
   * Renders the components
   */
  render(): JSX.Element {
    // tslint:disable-next-line: variable-name no-any
    const TagType = this.href === undefined ? 'button' : 'a' as any;
    const attrs = (TagType === 'button') ? { type: this.type } : { href: this.href };

    const classes = [];
    if (this.ghost) { classes.push('ghost'); }
    if (this.pill) { classes.push('pill'); }

    return (
      <TagType
        {...attrs}
        class={classes.join(' ')}
        disabled={this.disabled}
        onFocus={this.onFocus}
        onBlur={this.onBlur}>
        <slot />
      </TagType>
    );
  }
}
