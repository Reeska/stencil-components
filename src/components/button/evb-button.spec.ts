import { EvbButton } from './evb-button';

describe('EvbButton', () => {
  let component: EvbButton;

  beforeEach(() => component = new EvbButton());

  it('should render', () => {
    expect(component.render()).toMatchSnapshot();
  });

  it('should render with Props', () => {
    component.disabled = false;
    component.ghost = true;
    component.pill = true;
    component.type = 'submit';

    expect(component.render()).toMatchSnapshot();
  });

  it('should render as a reset button', () => {
    component.type = 'reset';

    expect(component.render()).toMatchSnapshot();
  });

  it('should render as a link', () => {
    component.href = '/assets/nice.gif';

    expect(component.render()).toMatchSnapshot();
  });

  it('should render as a link button', () => {
    component.type = 'link';

    expect(component.render()).toMatchSnapshot();
  });

});
