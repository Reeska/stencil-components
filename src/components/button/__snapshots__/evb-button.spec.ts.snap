// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvbButton should render 1`] = `
Object {
  "elm": undefined,
  "ishost": false,
  "vattrs": Object {
    "class": "",
    "disabled": false,
    "onBlur": [Function],
    "onFocus": [Function],
    "type": "button",
  },
  "vchildren": Array [
    Object {
      "elm": undefined,
      "ishost": false,
      "vattrs": null,
      "vchildren": null,
      "vkey": undefined,
      "vname": undefined,
      "vtag": "slot",
      "vtext": undefined,
    },
  ],
  "vkey": undefined,
  "vname": undefined,
  "vtag": "button",
  "vtext": undefined,
}
`;

exports[`EvbButton should render as a link 1`] = `
Object {
  "elm": undefined,
  "ishost": false,
  "vattrs": Object {
    "class": "",
    "disabled": false,
    "href": "/assets/nice.gif",
    "onBlur": [Function],
    "onFocus": [Function],
  },
  "vchildren": Array [
    Object {
      "elm": undefined,
      "ishost": false,
      "vattrs": null,
      "vchildren": null,
      "vkey": undefined,
      "vname": undefined,
      "vtag": "slot",
      "vtext": undefined,
    },
  ],
  "vkey": undefined,
  "vname": undefined,
  "vtag": "a",
  "vtext": undefined,
}
`;

exports[`EvbButton should render as a reset button 1`] = `
Object {
  "elm": undefined,
  "ishost": false,
  "vattrs": Object {
    "class": "",
    "disabled": false,
    "onBlur": [Function],
    "onFocus": [Function],
    "type": "reset",
  },
  "vchildren": Array [
    Object {
      "elm": undefined,
      "ishost": false,
      "vattrs": null,
      "vchildren": null,
      "vkey": undefined,
      "vname": undefined,
      "vtag": "slot",
      "vtext": undefined,
    },
  ],
  "vkey": undefined,
  "vname": undefined,
  "vtag": "button",
  "vtext": undefined,
}
`;

exports[`EvbButton should render with Props 1`] = `
Object {
  "elm": undefined,
  "ishost": false,
  "vattrs": Object {
    "class": "ghost pill",
    "disabled": false,
    "onBlur": [Function],
    "onFocus": [Function],
    "type": "submit",
  },
  "vchildren": Array [
    Object {
      "elm": undefined,
      "ishost": false,
      "vattrs": null,
      "vchildren": null,
      "vkey": undefined,
      "vname": undefined,
      "vtag": "slot",
      "vtext": undefined,
    },
  ],
  "vkey": undefined,
  "vname": undefined,
  "vtag": "button",
  "vtext": undefined,
}
`;
